### ISSUE TITLE
[python-package][dask] DaskLGBMClassifier.evals_result_ returns evals_result_ of one worker

### ISSUE THREAD

**Issue by szjunma** [microsoft/LightGBM#6963]
## Description
After training with `DaskLGBMClassifier` using multi-worker cluster, the `evals_result_` attribute returns evals_result_ of just one worker instead of aggregated results of all workers. The fitted model's `booster_.best_score` also only returns the score of the last iteration of one worker.

## Reproducible example
```
from sklearn.datasets import make_classification
import dask.dataframe as dd
from dask_ml.model_selection import train_test_split
import pandas as pd
import lightgbm as lgb
from lightgbm import DaskLGBMClassifier

## generate a dataset for classification
X, y = make_classification(
    n_samples=100_000, n_features=20, n_classes=2, random_state=42
)

df = pd.DataFrame(X, columns=[f"feature_{i}" for i in range(X.shape[1])])
df["target"] = y
dask_df = dd.from_pandas(df, npartitions=10)

train_df, eval_df = train_test_split(dask_df)

X_train = train_df.drop("target", axis=1)
y_train = train_df["target"]
X_eval = eval_df.drop("target", axis=1)
y_eval = eval_df["target"]

from dask.distributed import Client, LocalCluster

# Initialize Dask client
cluster = LocalCluster(n_workers=4, threads_per_worker=1, memory_limit="2GB")
client = Client(cluster)

# callback is not supported in DaskLGBMClassifier
# Only use this callback for logging eval metrics on the eval set on each worker
callback = lgb.log_evaluation(period=1)

# Train LightGBM Dask estimator
model = DaskLGBMClassifier(n_estimators=10, random_state=42, client=client)
model.fit(X_train, y_train, eval_set=[(X_eval, y_eval)], callbacks=[callback])

client.close()
```
Output (truncated):
...
[9]	valid_0's binary_logloss: 0.4072
[9]	valid_0's binary_logloss: 0.399198
[9]	valid_0's binary_logloss: 0.405525
[9]	valid_0's binary_logloss: 0.393885
[10]	valid_0's binary_logloss: 0.385598
[10]	valid_0's binary_logloss: 0.394278
[10]	valid_0's binary_logloss: 0.392088
[10]	valid_0's binary_logloss: 0.379982
[LightGBM] [Info] Finished linking network in 0.106994 seconds
[LightGBM] [Info] Finished linking network in 0.088812 seconds
[LightGBM] [Info] Finished linking network in 0.050092 seconds
[LightGBM] [Info] Finished linking network in 0.120033 seconds

```
model.evals_result_
```
Output:
{'valid_0': OrderedDict([('binary_logloss',
               [np.float64(0.6341578678872805),
                np.float64(0.5858938396397078),
                np.float64(0.5461053372128652),
                np.float64(0.5126605782706755),
                np.float64(0.48428744446358873),
                np.float64(0.4604012621581769),
                np.float64(0.4399457366016582),
                np.float64(0.4223975421284385),
                np.float64(0.4072001399359704),
                np.float64(0.39427754908759083)])])}

```
model.booster_.best_score
```
Output:
defaultdict(collections.OrderedDict,
            {'valid_0': OrderedDict([('binary_logloss',
                           np.float64(0.39427754908759083))])})
```
from sklearn.metrics import log_loss

pred_proba = model.predict_proba(X_eval).compute()
log_loss(y_eval.compute(), pred_proba[:, 1]) # this should be weighted avg across all the workers
```
Output:
0.3875167385783791

## Environment info

LightGBM version or commit hash: Commit 195c26f

Command(s) you used to install LightGBM

```shell
git clone --recursive https://github.com/microsoft/LightGBM
cd LightGBM
cmake -B build -S .
cmake --build build -j4

build-python.sh install --precompile
```

```
pip list
```
Package                 Version
----------------------- -----------
appnope                 0.1.4
asttokens               3.0.0
build                   1.2.2.post1
click                   8.2.1
cloudpickle             3.1.1
comm                    0.2.2
dask                    2025.5.1
dask-glm                0.3.2
dask-ml                 2025.1.0
debugpy                 1.8.11
decorator               5.2.1
distributed             2025.5.1
exceptiongroup          1.3.0
executing               2.2.0
fsspec                  2025.5.1
importlib_metadata      8.7.0
ipykernel               6.29.5
ipython                 9.4.0
ipython_pygments_lexers 1.1.1
jedi                    0.19.2
Jinja2                  3.1.6
joblib                  1.5.1
jupyter_client          8.6.3
jupyter_core            5.8.1
lightgbm                4.6.0.99
llvmlite                0.44.0
locket                  1.0.0
MarkupSafe              3.0.2
matplotlib-inline       0.1.7
msgpack                 1.1.1
multipledispatch        1.0.0
nest_asyncio            1.6.0
numba                   0.61.2
numpy                   2.2.6
packaging               25.0
pandas                  2.3.0
parso                   0.8.4
partd                   1.4.2
pexpect                 4.9.0
pickleshare             0.7.5
pip                     25.1
platformdirs            4.3.8
prompt_toolkit          3.0.51
psutil                  5.9.0
ptyprocess              0.7.0
pure_eval               0.2.3
pyarrow                 20.0.0
Pygments                2.19.2
pyproject_hooks         1.2.0
python-dateutil         2.9.0.post0
pytz                    2025.2
PyYAML                  6.0.2
pyzmq                   26.2.0
scikit-learn            1.7.0
scipy                   1.16.0
setuptools              78.1.1
six                     1.17.0
sortedcontainers        2.4.0
sparse                  0.17.0
stack_data              0.6.3
tblib                   3.1.0
threadpoolctl           3.6.0
toolz                   1.0.0
tornado                 6.5.1
traitlets               5.14.3
typing_extensions       4.14.1
tzdata                  2025.2
urllib3                 2.5.0
wcwidth                 0.2.13
wheel                   0.45.1
zict                    3.0.0
zipp                    3.23.0

## Additional Comments

When doing `booster.eval_valid(feval)` as [here](https://github.com/ffineis/LightGBM/blob/master/python-package/lightgbm/engine.py#L256), shouldn't the worker talk to other workers and do aggregation/sync in master worker, just as `booster.update(fobj=fobj)`? However I couldn't find such in lightgbm's built-in distributed training code.

**Comment by szjunma** [comment]
Re-running the same example on the PR #7290 branch now produces:

```
...
[9]	valid_0's binary_logloss: 0.398119
[9]	valid_0's binary_logloss: 0.398119
[9]	valid_0's binary_logloss: 0.398119
[9]	valid_0's binary_logloss: 0.398119
[10]	valid_0's binary_logloss: 0.384417
[10]	valid_0's binary_logloss: 0.384417
[10]	valid_0's binary_logloss: 0.384417
[10]	valid_0's binary_logloss: 0.384417
```

All 4 workers now report the same globally-aggregated value at each iteration (no more per-worker drift). The end result:

```python
model.evals_result_
# {'valid_0': OrderedDict([('binary_logloss',
#               ...,
#               np.float64(0.384417128689068)])])}

model.booster_.best_score
# defaultdict(..., {'valid_0': OrderedDict({'binary_logloss': np.float64(0.384417128689068)})})

from sklearn.metrics import log_loss
pred_proba = model.predict_proba(X_eval).compute()
log_loss(y_eval.compute(), pred_proba[:, 1])
# 0.384417128689068
```

Reported metric now matches the ground-truth global `log_loss` exactly. This covers all built-in additive metrics (binary_logloss, multi_logloss, l2, rmse, cross_entropy, etc.). Non-additive metrics (AUC, NDCG, MAP, AveragePrecision, auc_mu) are handled in a follow-up branch with different aggregation strategies.
